### DISCUSSION TITLE
libtorch jit forward error

### DISCUSSION THREAD

**Discussion by kli017** [snakers4/silero-vad#154] [Q&A] [answered]
## ❓ Questions and Help
Hello I was implement the new vad model with libtorch jit. The init and forward code as bellow
`
static float ComputeConfidence(const std::vector<torch::jit::IValue> &inputs, const int sr) {
    torch::NoGradGuard no_grad;
    auto output = model->forward(inputs).toTensor();
    return output[0][1].item().toFloat();
}

bool InitTorchVadModel(const TorchVadOptions &opts) {
    model = std::make_shared<TorchVadModule>(
        std::move(torch::jit::load(opts.model_path)));
    model->eval();
    model->to(at::kCPU);
    at::set_num_threads(opts.num_thread);
    return model != nullptr;
}
`
while make the project I got the error as below, which said that the forward function need a parameter named sr.
terminate called after throwing an instance of 'c10::Error'
  what():  forward() is missing value for argument 'sr'. Declaration: forward(__torch__.HehRecModel self, Tensor x, int sr) -> (Tensor)
Exception raised from checkAndNormalizeInputs at ../aten/src/ATen/core/function_schema_inl.h:239 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x69 (0x7f2e48c5d9f9 in /root/kli/libtorch_vad/fc_base/libtorch-src/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xd2 (0x7f2e48c5a012 in /root/kli/libtorch_vad/fc_base/libtorch-src/lib/libc10.so)
frame #2: <unknown function> + 0xf10708 (0x7f2e368db708 in /root/kli/libtorch_vad/fc_base/libtorch-src/lib/libtorch_cpu.so)
frame #3: torch::jit::GraphFunction::operator()(std::vector<c10::IValue, std::allocator<c10::IValue> >, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, c10::IValue> > > const&) + 0x2d (0x7f2e39209b2d in /root/kli/libtorch_vad/fc_base/libtorch-src/lib/libtorch_cpu.so)
frame #4: torch::jit::Method::operator()(std::vector<c10::IValue, std::allocator<c10::IValue> >, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, c10::IValue, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, c10::IValue> > > const&) + 0x16a (0x7f2e39217f8a in /root/kli/libtorch_vad/fc_base/libtorch-src/lib/libtorch_cpu.so)
frame #5: torch::jit::Module::forward(std::vector<c10::IValue, std::allocator<c10::IValue> >) + 0xff (0x7f2e48874e3b in liblibtorch_vad.so)
frame #6: <unknown function> + 0x3184d (0x7f2e4886f84d in liblibtorch_vad.so)
frame #7: voicecomm::vad::TorchVad::IsSpeech(std::vector<float, std::allocator<float> > const&) + 0x13f (0x7f2e4886fd4d in liblibtorch_vad.so)
frame #8: voicecomm::vad::TorchVad::IsSpeech(std::vector<char, std::allocator<char> > const&) + 0x46 (0x7f2e4886fbc6 in liblibtorch_vad.so)
frame #9: <unknown function> + 0x2b47 (0x5577003e9b47 in ./libtorch_vad_main)
frame #10: __libc_start_main + 0xe7 (0x7f2e3505abf7 in /lib/x86_64-linux-gnu/libc.so.6)
frame #11: <unknown function> + 0x286a (0x5577003e986a in ./libtorch_vad_main)

**Comment by snakers4** [comment, marked as answer]
> while make the project I got the error as below, which said that the forward function need a parameter named sr.

Yes, this is correct

https://github.com/snakers4/silero-vad/blob/76687cbe25ffdf992ad824a36bfe73f6ae1afe72/utils_vad.py#L205-L211

**Reply by snakers4** [reply]
`sr` is just sampling rate
The JIT model accepts either `8000` or `16000` 
For higher sampling rates this hack can be applied

https://github.com/snakers4/silero-vad/blob/76687cbe25ffdf992ad824a36bfe73f6ae1afe72/utils_vad.py#L185-L191

**Reply by kli017** [reply]
Thanks for the reply, yes I am using 16000 wav data. I also tried forward with added the sr parameter but got another error. Is it beacuse the version of libtorch?

/root/kli/libtorch_vad/src/torch_vad.cc: In function ‘float ComputeConfidence(const std::vector<c10::IValue>&, int)’:
/root/kli/libtorch_vad/src/torch_vad.cc:39:44: error: no matching function for call to ‘torch::jit::Module::forward(const std::vector<c10::IValue>&, const int&)’
     auto output = model->forward(inputs, sr).toTensor();
                                            ^
In file included from /root/kli/libtorch_vad/fc_base/libtorch-src/include/torch/csrc/jit/serialization/import.h:5:0,
                 from /root/kli/libtorch_vad/fc_base/libtorch-src/include/torch/script.h:8,
                 from /root/kli/libtorch_vad/src/torch_vad.cc:3:
/root/kli/libtorch_vad/fc_base/libtorch-src/include/torch/csrc/jit/api/module.h:113:10: note: candidate: c10::IValue torch::jit::Module::forward(std::vector<c10::IValue>)
   IValue forward(std::vector<IValue> inputs) {
          ^~~~~~~
/root/kli/libtorch_vad/fc_base/libtorch-src/include/torch/csrc/jit/api/module.h:113:10: note:   candidate expects 1 argument, 2 provided
CMakeFiles/libtorch_vad.dir/build.make:62: recipe for target 'CMakeFiles/libtorch_vad.dir/src/torch_vad.cc.o' failed
make[2]: *** [CMakeFiles/libtorch_vad.dir/src/torch_vad.cc.o] Error 1
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/libtorch_vad.dir/all' failed
make[1]: *** [CMakeFiles/libtorch_vad.dir/all] Error 2
Makefile:162: recipe for target 'all' failed
make: *** [all] Error 2

**Reply by snakers4** [reply]
> error: no matching function for call to ‘torch::jit::Module::forward(const std::vectorc10::IValue&, const int&)
> note: candidate expects 1 argument, 2 provided

To be honest this is a bit cryptic to me, looks like to me that it complains about actually having 2 parameters instead of one
Maybe there some limitations in `libtorch` or `model.forward` has to be invoked differently if it has several params?

**Reply by kli017** [reply]
Yes, The previous jit model which only need inputdata works well. no idea how to deal with the error.

**Reply by snakers4** [reply]
Maybe this would help - https://github.com/pytorch/pytorch/issues/18337#issuecomment-475709658

**Reply by kli017** [reply]
Thanks! So should I convert the sr from int to a tensor and put it to the inputs? I tried so but got **what():  forward() Expected a value of type 'int' for argument 'sr' but instead found type 'Tensor'.**
Seems the int sr should be an independet paramter..

**Reply by snakers4** [reply]
They mention creating a special `IValue` object
I am not familiar with C++ API, but looks like a dictionary to me
